### PR TITLE
[NEW] OAuth groups to channels mapping

### PR DESCRIPTION
**Pull request by arminfelder**
adds the option to map oauth groups to rocketchat channels, in a similar way it is already possible for LDAP

#Closes https://github.com/RocketChat/feature-requests/issues/360

### CODE REVIEW

**Comment by pierreozoux** [comment]
@pierre-lehnen-rc @sampaiodiego @arminfelder any news to get this PR moving?

Any blocker?

Thanks a lot @arminfelder, this is a useful feature :)

File: app/custom-oauth/server/oauth_helpers.js [hunk 1]
```diff
@@ -40,3 +43,31 @@ export function updateRolesFromSSO(user, identity, roleClaimName) {
...
+		for (const ssoGroup in channelsMap) {
+			if (typeof ssoGroup === 'string') {
+				let channels = channelsMap[ssoGroup];
+				if (!Array.isArray(channels)) {
+					channels = [channels];
+				}
+				for (const channel of channels) {
+					let room = Rooms.findOneByNonValidatedName(channel);
+					if (!room) {
+						room = createRoom('c', channel, channelsAdmin, [], false);
```
Comment (geekgonecrazy) [inline app/custom-oauth/server/oauth_helpers.js:60]: Is this save to be doing here?  When does this run?  Isn't this running every time an oauth request comes in?

Comment (arminfelder) [reply app/custom-oauth/server/oauth_helpers.js:60]: it should run on every oauth request, to in case apply updates

**Comment by rasos** [comment]
We are using this patch in a production environment - it works like a charm. Please integrate upstream.

**Comment by ImaCrea** [comment]
Any feedback on pushing it upstream please?

File: app/lib/server/startup/oAuthServicesUpdate.js [hunk 2]
```diff
@@ -51,7 +51,11 @@ function _OAuthServicesUpdate() {
 				data.nameField = settings.get(`${ service.key }-name_field`);
 				data.avatarField = settings.get(`${ service.key }-avatar_field`);
 				data.rolesClaim = settings.get(`${ service.key }-roles_claim`);
+				data.groupsClaim = settings.get(`${ service.key }-groups_claim`);
+				data.channelsMap = settings.get(`${ service.key }-groups_channel_map`);
+				data.channelsAdmin = settings.get(`${ service.key }--channels_admin`);
```
Comment (pierre-lehnen-rc) [inline app/lib/server/startup/oAuthServicesUpdate.js:56]:
```suggestion
				data.channelsAdmin = settings.get(`${ service.key }-channels_admin`);
```

File: app/custom-oauth/server/oauth_helpers.js [hunk 3]
```diff
@@ -40,3 +43,31 @@ export function updateRolesFromSSO(user, identity, roleClaimName) {
...
+						if (!room || !room.rid) {
+							logger.error(`could not create channel ${ channel }`);
+							return;
+						}
+					}
+					if (groupsFromSSO.includes(ssoGroup)) {
```
Comment (pierre-lehnen-rc) [inline app/custom-oauth/server/oauth_helpers.js:66]:
```suggestion
					if (Array.isArray(groupsFromSSO) && groupsFromSSO.includes(ssoGroup)) {
```

**Review by pierre-lehnen-rc — CHANGES_REQUESTED** [review]

**Review by pierre-lehnen-rc — APPROVED** [review]

**Pull request merged**